lib/active_merchant/billing/gateways/quickbooks.rb

```diff
@@ -0,0 +1,312 @@
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((realm=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_consumer_key=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_nonce=\")\w+), '\1[FILTERED]').
+          gsub(%r((oauth_signature=\")[a-zA-Z%0-9]+), '\1[FILTERED]').
+          gsub(%r((oauth_token=\")\w+), '\1[FILTERED]').
+          gsub(%r((\"card\":{\"number\":\")\d+), '\1[FILTERED]').
+          gsub(%r((\"cvc\":\")\d+), '\1[FILTERED]')
+      end
+
+      private
+
+      def add_charge_data(post, payment, options = {})
+        add_payment(post, payment, options)
+        add_address(post, options)
+        add_context(options[:context]) if options[:context]
+      end
+
+      def add_address(post, options)
+        return unless post[:card] && post[:card].kind_of?(Hash)
+
+        card_address = {}
+        if address = options[:billing_address] || options[:address]
+          card_address[:streetAddress] = address[:address1]
+          card_address[:city] = address[:city]
+          card_address[:region] = address[:state] || address[:region]
+          card_address[:country] = address[:country]
+          card_address[:postalCode] = address[:zip] if address[:zip]
+        end
+        post[:card][:address] = card_address
+      end
+
+      def add_amount(post, money, options = {})
+        currency = options[:currency] || currency(money)
+        post[:amount] = localized_amount(money, currency)
+        post[:currency] = currency.upcase
+      end
+
+      def add_payment(post, payment, options = {})
+        add_creditcard(post, payment, options)
+      end
+
+      def add_creditcard(post, creditcard, options = {})
+        card = {}
+        card[:number] = creditcard.number
+        card[:expMonth] = "%02d" % creditcard.month
+        card[:expYear] = creditcard.year
+        card[:cvc] = creditcard.verification_value if creditcard.verification_value?
+        card[:name] = creditcard.name if creditcard.name
+        card[:commercialCardCode] = options[:card_code] if options[:card_code]
+
+        post[:card] = card
+      end
+
+      def add_context(post, context)
+        payment_context = {}
+        payment_context[:tax] = context[:tax] if context[:tax]
+        payment_context[:recurring] = context[:recurring] if context[:recurring]
+
+        post[:context] = payment_context
+      end
+
+      def auth_object_from(authorization)
+        uri = "#{ENDPOINT}/#{authorization}"
+        commit(uri, {}, :get)
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def commit(uri, body = {}, method = :post)
+        endpoint = gateway_url + uri
+        begin
+          response = case method
+          when :post
+            ssl_post(endpoint, post_data(body), headers(method: :post, uri: endpoint))
+          when :get
+            ssl_request(:get, endpoint, nil, headers(method: :get, uri: endpoint))
+          end
+        rescue ResponseError => e
```

Comment:
I'm not sure we want to rescue `ResponseError`, since the caller probably expects to handle it themselves (I know Spreedly does).